stubby: some fixes, cleanup, and improvements

guidosarducci opened the pull request:
Maintainer: @iamperson347 
Compile tested: mips_24kc, DIR-835, LEDE 17.01.5
Run tested: mips_24kc, DIR-835, LEDE 17.01.5

#### Description:
This includes several updates for issues noticed over the last few weeks:
- Makefile cleanup and fix a broken `conffiles` section (also seen in #6682)
- add Cloudflare servers and corresponding SPKI
- set EDNS client-subnet consistent with upstream for improved privacy
- depend on `ca-certificates` package needed for TLS verification
- drop unneeded `procd` "core" resource limit

**Note**: The current breakdown of commits is to ease review. Please squash the commit sequence after discussion/review and before any final merge.

karlp commented on net/stubby/files/README.md:
```diff
@@ -20,10 +20,10 @@ This package has some modifications that makes it differ from the default upstre
 The value of "edns_client_subnet_private" is '1' in the upstream default config. This informs the upstream resolver to NOT forward your connection's IP to any other upstream servers. This is good for privacy, but could result in sub-optimal routing to CDNs, etc.
 
-To give a more "comparable" DNS experience similar to google/opendns, this package disables this option.
+We retain the upstream value since privacy is a key user motivation for using DNS-over-TLS, but note users encountering poor routing may consider changing it.
```
This type of change should really be done standalone IMO, not buried inside a "fixes and cleanups" type patch.  However, I'm not a user of thisw package

iamperson347 replied:
I agree. However, this would more closely match upstream. When I originally published this package, I was going with what I believed would be a good overall default config for the largest number of users. However, we could easily argue it should match upstream and users tweak as desired.

karlp commented:
"stubby: rearrange Makefile for clarity" is just noise.

are you _sure_ that the whitespace is the cause of your problems with conffiles?  that seems very wrong, and if it is the case, it's a problem in core, not a problem in the defintioin of the config files.

for "stubby: add SPKI pinset for Cloudflare cert" could you explain _why_ you're doing it, not just that you've done it?  What happens without it? What will happen after that date?

iamperson347 commented on net/stubby/files/stubby.yml:
```diff
@@ -24,6 +24,14 @@ upstream_recursive_servers:
 # # Cloudflare IPv6
   - address_data: 2606:4700:4700::1111
     tls_auth_name: "cloudflare-dns.com"
+    tls_pubkey_pinset:
```
From what I have read, it is not currently recommended to do this with the cloudflare service.

https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Public+Resolvers

iamperson347 commented on net/stubby/files/stubby.yml:
```diff
@@ -33,6 +41,14 @@ upstream_recursive_servers:
 # # Cloudflare servers
   - address_data: 1.1.1.1
     tls_auth_name: "cloudflare-dns.com"
+    tls_pubkey_pinset:
```
From what I have read, it is not currently recommended to do this with the cloudflare service.

https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Public+Resolvers

iamperson347 replied:
From what I have read, it is not currently recommended to do this with the cloudflare service.

https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Public+Resolvers

jonathanunderwood commented:
The ca-certificates dependency change has been split out into PR #7030. It'd be great if we could start getting some of this merged.

jonathanunderwood commented:
The stubby: use EDNS client-subnet privacy by default commit has been split out into PR #7031 as requested.

jonathanunderwood commented:
The commits in this PR have now been split out into a series of separate PRs: #7030 #7031 #7032 #7033 #7034. Hopefully this will make review and merging more straightforward.

guidosarducci commented:
@jonathanunderwood  Just what do you think you're doing?!  I could understand if this PR was several months old and you'd already made several unsuccessful attempts to contact me... but none of this happened.

Instead, you simply hijacked my PR, and then split out each commit into your own PR, creating a lot of pointless repo churn for anyone reviewing or following. Then you copy my "`Signed-off-by:`" line at the top of each of your PRs, in one case fronting for your own changes. This is *completely unacceptable*. Please delete your multiple PR "clones" so I can continue on mine without this unwelcome confusion.

CC: @dibdot

jonathanunderwood commented:
@guidosarducci wow, there's really no need for that attitude. 

1. If you look carefully at the other PRs, you'll see that I have literally split out *your* commits (using git cherry-pick). Because they are exactly your commits (which you can verify by comparing the checksum commits), of course they contain your signed-off by in the commit message.

2. The reason I split them out into separate PRs is because that is what the stubby package maintainer had requested in order to make review easier, and there'd been no activity from you since 7th August. I have work locally I'd like to submit that builds on your work, so I felt it was a good use of time to speed up the review and acceptance of that parts of your PR that were already acceptable to the package owner.

3. As an example, since the package owner had rejected the SKPI patch, it wasn't possible to simply merge the PR in its current form. Splitting out the PRs into single feature PRs was done to help review and simplify the process.

I am sorry, but I don't really see what the problem is here.

jonathanunderwood commented:
Aside:

> Then you copy my "Signed-off-by:" line at the top of each of your PRs, in one case fronting for your own changes.

When you create a PR, the template pulls the signed-off from the commits. In order to ensure that it was clear that you, and not me, were the author, I simply left the signed-off in the PR (to be consistent with the signedd-off in the commits).

I have no idea what you're talking about with "in one case fronting for your own changes". Evidence, please, or retract that accusation.

jonathanunderwood commented:
For reference:

1. @karlp requested the EDNS change be sumbitted via a separate PR
2. @iamperson347 has indicated he's uncertain about merging the SPKI change.

Both of those need addressing.

guidosarducci commented:
@iamperson347 
Hi David,
Regarding SPKI, the [3rd party information](https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Public+Resolvers) you refer to claiming that Cloudflare do not publish or recommend using SPKI pins is just plain wrong (perhaps out-of-date?). Cloudflare's own [configuration guide](https://developers.cloudflare.com/1.1.1.1/dns-over-tls/) in fact describes how to use SPKI and publishes their SHA256 pin.

The lack of pin set data was previously raised in #5552 by @EnergyMouse , and Cloudflare SPKI is commonly used by stubby users as discussed in various online fora. I've now updated my SPKI commit to include some of the above clarification about the official Cloudflare position. There are no other objections I'm aware of so let's get this merged.

I also noticed in #5552 you justified disabling the `round_robin_upstreams` setting (contrary to upstream) in order to force IPv6 usage first. I've not heard of others doing this myself. Is it still your position, or are you open to using the upstream (enabled) setting?

@dibdot I bumped PKG_RELEASE  as requested.
>@iamperson347 as requested, the ca-certificates dependency has been submitted & merged via separate PR. Is it OK to merge the remaining PR?

Aside from my SPKI commit above, all the others have been approved by @iamperson347  I believe (either here or as highjacked copies) but I'll let him confirm.

Regarding the `ca-certificates` dependency, I dropped this commit from my PR branch since OpenWrt `master` is shifting to dependencies on `ca-bundle`, whose smaller size is more appropriate for constrained embedded platforms. There's some discussion [here](https://github.com/openwrt/packages/issues/6682#issuecomment-412508423), and what seemed like a plan to do the same for `stubby`. Instead, https://github.com/openwrt/packages/pull/7030 was merged, taking us a step backwards by enforcing a `ca-certificates` dependency. We should really consider reverting this in favour of `ca-bundle`.

Thanks!

EricLuehrsen commented:
Commits d5b0c46ece4bc494843ff99758704122684b864, 8f241854ed89d9fb70c2a10955af7041900e83f2 and 4819fc5e6e63551444af479054deb1556353b915 should be squashed and moved to the last commit.

Commits 0425d9198a41c8c774e599195b59b3b931b2f9b4 and 8b2de594de0219681ba9630b8390738a1afb7e4e should have respective README.md changes with the options that changed, or because they are both default options just squash.

Commit 7a1cfd43e956235010e321aa4c2cbcf0d12a56aa and PR #7029 should be one. Really `stubby.init` could use some care. Copy with pride from `dnsmasq`, `unbound`, or `adblock`.

guidosarducci commented:
@EricLuehrsen 
Hi Eric,
Thanks for the comments.  I noted in the PR intro that there's further squashing needed before a merge. The current breakout of commits is purely for review, and once the maintainer @iamperson347 confirms one last commit, I'll squash and ask @dibdot  to merge.

>Commits 0425d91 and 8b2de59 should have respective README.md changes with the options that changed, or because they are both default options just squash.

There is already an upstream configuration reference included in the package at `/etc/stubby/stubby.yml.default`, which provides background on the various options. The `README.md` has been used to capture key configuration rationales as well as package-specific changes. Commit 0425d91 includes `README.md` updates, and notes already exist for 8b2de59 on use of Cloudflare.

>Really stubby.init could use some care. Copy with pride from dnsmasq, unbound, or adblock.

Yeah, I'm familiar with writing/fixing procd-based init files, but `stubby.init` is already pretty bare-bones.

Yes, there's minor stuff the maintainer @iamperson347 might comment on: Why need an explicit `STOP=51` at all? Is there a reason for `START=50` being out of line with the similar `dnsmasq`? Why the redundant default parameters to `procd_set_param respawn`?

I know most users would have preferred having a UCI framework for example, but really that's not in scope for this PR. Are there specific, serious problems you see?

>Commit 7a1cfd4 and PR #7029 should be one.

I disagree. IIRC that PR isn't really appropriate; it tries to effectively implement a boot delay with a kludgey use of interface triggers. The root problem seems to be that `stubby` won't retry retrieval of its DNSSEC trust anchors after an initial failure, which is an issue best raised and fixed upstream, then imported e.g. via patch.

Thanks!

dibdot commented:
Let's start merging these pending stubby PRs for better OpenWrt integration. Thanks for your patience.
